hudi-client/hudi-client-common/src/main/java/org/apache/hudi/io/FileGroupReaderBasedMergeHandle.java

```diff
@@ -198,8 +270,12 @@ public void doMerge() {
           }
           // Writes the record
           try {
-            writeToFile(record.getKey(), record, writeSchemaWithMetaFields,
-                config.getPayloadConfig().getProps(), preserveMetadata);
+            // For Compaction operations, the preserveMetadata flag is always true as we want to preserve the existing record metadata.
+            // For other updates, we only want to preserve the metadata if the record is not being modified by this update. If the record already exists in the base file and is not updated,
+            // the operation will be null. Records that are being updated or records being added to the file group for the first time will have an operation set and must generate new metadata.
+            boolean shouldPreserveRecordMetadata = preserveMetadata || record.getOperation() == null;
+            Schema recordSchema = shouldPreserveRecordMetadata ? writeSchemaWithMetaFields : writeSchema;
+            writeToFile(record.getKey(), record, recordSchema, config.getPayloadConfig().getProps(), shouldPreserveRecordMetadata);
```

Comment:
Currently we collect SI and RLI no matter wheter the record has been written successfully, this could incur inconsistencies.

Reply:
Create this JIRA [ticket](https://issues.apache.org/jira/browse/HUDI-9724) to track and will take it up in the next day or so